autolab/Autolab
# Display Grace Day usage on submission history table, improve management of assessment penalty settings

## Summary

### Summary generated by Reviewpad on 01 Nov 23 03:23 UTC
This pull request includes several changes across multiple files. Here's a summary of the changes:

- In the file `_edit_penalties.html.erb`, changes were made to text fields, labels, and checkboxes related to assessment penalties. The changes aim to improve user experience and provide more flexibility in setting penalties for assessments.
- In the file `assessments_controller.rb`, modifications were made to the `edit` and `update` actions to handle penalty-related parameters and provide warning messages for instructors. These changes enhance the functionality and flexibility of the assessment controller.
- The file `courses/_courseFields.html.erb` includes modifications to the placeholder values of certain input fields.
- Changes in the `FormBuilderWithDateTimeInput` class consolidate and improve the arguments and options passed to various methods.
- The file diff related to the Penalties tab in the assessment editor introduces event listeners for checkboxes, logic to enable or disable input fields, and triggering change events on page load.
- Changes in the `ScoreAdjustment` model's `to_s` method enhance consistency and formatting.
- The `app/views/assessments/_submission_history_table.html.erb` file modifies a label in the submission history table.
- CSS selectors in another file now exclude elements with a specific class and target additional elements, adding a bottom padding.
- The `edit.html.erb` file in the `app/views/assessments` directory removes error message handling code.
- A new file `config/initializers/core_ext.rb` is being added, defining class extensions for string-to-boolean conversions.
- Changes in the `app/views/assessments/_submission_history_row.html.erb` file enhance the display of submission details.

These changes improve functionality, user experience, and consistency in different parts of the codebase.


## Description

- Display grace days used next to number of late days
- Add warnings when late days are allowed but settings do not make sense (e.g. grace days can't be used and/or penalty late days are "free")
- Add checkboxes on penalty tab to easily use course defaults / allow unlimited submissions / allow unlimited grace days

## Motivation and Context



Closes #1989

## How Has This Been Tested?



Make a late submission on an assessment that allows the use of grace days. Observe the correct number of grace days used displayed.

![Screenshot 2023-10-03 at 16 23 02](https://github.com/autolab/Autolab/assets/9074856/0ff1c52d-bdab-401c-bdad-4c3fd10b1353)

Allow late submissions on assessment (end at > due at). Set late penalty to 0 and/or max grace days to 0, see warning(s). Also check that they display correctly when course defaults are used for these values.

![Screenshot 2023-10-06 at 01 15 52](https://github.com/autolab/Autolab/assets/9074856/fc1ee167-7a8e-48b4-b4ef-050e88d7b7bf)

- Set values for max submissions, max grace days, late penalty, version threshold, version penalty -> ensure it saves successfully
- Check all the checkboxes and save
  - max submissions -> -1
  - max grace days -> (blank)
  - late penalty -> (blank)
  - version threshold -> -1
  - version penalty -> (blank)
- Verify that the above settings take effect
  - make a student account submit once, a few days late
  - they should see unlimited submissions left
  - they should see all their grace days being used (e.g. 3)
  - can test late penalty by toggling between 0 (assessment-specific), and 1 (course default)
  - can test version threshold / penalty in a similar manner

![Screenshot 2023-10-12 at 22 37 07](https://github.com/autolab/Autolab/assets/9074856/fece1920-4ac5-4d73-bdfa-b26be2775d92)
![Screenshot 2023-10-12 at 22 37 13](https://github.com/autolab/Autolab/assets/9074856/1262cce5-4a23-4eef-b116-ffbf29e701ad)

## Types of changes


## Checklist:
